### DISCUSSION TITLE
Can't navigate when setBackgroundMessageHandler is triggered

### DISCUSSION THREAD

**Discussion by youssefhaitham-callvita** [invertase/react-native-firebase#5509] [Q&A]
Hello,

I am trying to navigate to a specific page when a user opens a notification and boots the application. In the React native firebase documentation it used setBackgroundMessageHandler to handle the notification sent but it handles it outside the whole app in the index.js file. 

```
import {AppRegistry} from 'react-native';
import App from './App';
import {name as appName} from './app.json';
import messaging from '@react-native-firebase/messaging';
import React from 'react';

messaging().setBackgroundMessageHandler(async remoteMessage => {
 console.log('Message handled in the background!', remoteMessage);
});

function HeadlessCheck({ isHeadless }) {
    if (isHeadless) {
      // App has been launched in the background by iOS, ignore
      return null;
    }
  
    return <App />;
  }

AppRegistry.registerComponent(appName, () => HeadlessCheck);
```

So my question is, how can I navigate to a certain module if the app throw the setBackgroundMessageHandler isn't registered yet?

**Comment by mikehardy** [comment]
Integrate some form of persistent storage, shared between components of the app.
Save the fact you have some place to go, and enqueue where that should be, in this persistent storage in the background handler.
During bootup, at an appropriate time, check to see if you have some place to go, and if you do, dequeue that location and go to it

**Reply by youssefhaitham-callvita** [reply]
Thank you so much, I will try it now hope it works.

**Reply by mikehardy** [reply]
For what it's worth, this is the typical use case for redux or resub - that is, "persistent" can be in memory but persisted between different components - it does not have to be async-storage module or things like that although those also work - good luck!

**Comment by JoseChavez98** [comment]
Hi! @mikehardy , I'm using async storage to build the queue, It's working, but I have a problem. 
When I receive a notification and ignore it and open my app, the **setBackgroundMessageHandler** is triggering. 
The expected behavior is that the **setBackgroundMessageHandler** only triggers when clicking the notification.
Right now, my logic for creating the queue is inside the **setBackgroundMessageHandler**.
this is from 'quit' state.

```
messaging().setBackgroundMessageHandler(async remoteMessage => {
  var processedMessage = CometChat.CometChatHelper.processMessage(
    JSON.parse(remoteMessage.data.message)
  );
  const redirectObj = {
    stack: 'ChatStack',
    screen: 'ChatMessages',
    params: {
      item: { ...processedMessage.sender },
      tab: 'conversations',
      actionGenerated: actionType => {}
    }
  };
  await setItem(REDIRECT, JSON.stringify(redirectObj));
});
```

**Reply by mikehardy** [reply]
Hey @JoseChavez98 - without seeing the full contents of the JSON you send via the FCM REST API for testing (minus the device token for security reasons), I can't speculate. If you could post that I may be able to deduce why you see what you see.

**Reply by JoseChavez98** [reply]
@mikehardy thanks for the response!!

It is a special JSON that Cometchat generates, but the important thing here is that if I'm right in saying that the **setBackgroundMessageHandler** should only trigger when notification is clicked?. Also I saw other methods like ***onNotificationOpenedApp*** , when should I use this ones instead of **setBackgroundMessageHandler**.
The [documentation](https://rnfirebase.io/messaging/usage#background--quit-state-messages) says that for quit and background state, **setBackgroundMessageHandler** must be enough.
